lib/net/nforceif/src/sys_arch.c

```diff
@@ -91,209 +64,176 @@ sys_init(void)
 {
 }
 
-#if 0
-u32_t
-sys_jiffies(void)
-{
-    struct timespec ts;
-
-    clock_gettime(CLOCK_MONOTONIC, &ts);
-    return ts.tv_sec * 1000000000L + ts.tv_nsec;
-}
-#endif
-
-#if !NO_SYS
-
 sys_thread_t
 sys_thread_new(const char *name, lwip_thread_fn function, void *arg, int stacksize, int prio)
 {
     struct sys_thread *st = NULL;
     LWIP_UNUSED_ARG(name);
     LWIP_UNUSED_ARG(prio);
 
-    st = (struct sys_thread *)malloc(sizeof(struct sys_thread));
-
-    if (NULL == st) {
-        abort();
-    }
-
-    st->handle = CreateThread(NULL, stacksize, (void *)function, arg, 0, NULL);
-
-    return st;
+    return CreateThread(NULL, stacksize, (void *)function, arg, 0, NULL);
 }
 
 err_t
-sys_mbox_new(struct sys_mbox **mb, int size)
+sys_mbox_new(sys_mbox_t *mb, int size)
 {
-    struct sys_mbox *mbox;
     LWIP_UNUSED_ARG(size);
 
-    mbox = (struct sys_mbox *)malloc(sizeof(struct sys_mbox));
-    if (mbox == NULL) {
-        return ERR_MEM;
-    }
-    mbox->first = mbox->last = 0;
-    mbox->read_sem = sys_sem_new_internal(0);
-    mbox->write_sem = sys_sem_new_internal(SYS_MBOX_SIZE);
-    mbox->mutex = sys_sem_new_internal(1);
+    mb->first = mb->last = 0;
+    sys_sem_new(&mb->read_sem, 0);
+    sys_sem_new(&mb->write_sem, SYS_MBOX_SIZE);
+    mb->valid = 1;
 
     SYS_STATS_INC_USED(mbox);
-    *mb = mbox;
     return ERR_OK;
 }
 
 void
-sys_mbox_free(struct sys_mbox **mb)
+sys_mbox_free(sys_mbox_t *mb)
 {
-    if ((mb != NULL) && (*mb != SYS_MBOX_NULL)) {
-        struct sys_mbox *mbox = *mb;
+    if (mb != NULL) {
         SYS_STATS_DEC(mbox.used);
-        sys_arch_sem_wait(&mbox->mutex, 0);
 
-        sys_sem_free_internal(mbox->read_sem);
-        sys_sem_free_internal(mbox->write_sem);
-        sys_sem_free_internal(mbox->mutex);
-        /*  LWIP_DEBUGF("sys_mbox_free: mbox 0x%lx\n", mbox); */
-        free(mbox);
+        sys_sem_free(&mb->read_sem);
+        sys_sem_free(&mb->write_sem);
+        LWIP_DEBUGF(SYS_DEBUG, ("sys_mbox_free: mbox 0x%p\n", (void *)mb));
     }
 }
 
 err_t
-sys_mbox_trypost(struct sys_mbox **mb, void *msg)
+sys_mbox_trypost(sys_mbox_t *mb, void *msg)
 {
     u8_t first;
-    struct sys_mbox *mbox;
-    LWIP_ASSERT("invalid mbox", (mb != NULL) && (*mb != NULL));
-    mbox = *mb;
+    LWIP_ASSERT("invalid mbox", mb != NULL);
 
     LWIP_DEBUGF(SYS_DEBUG, ("sys_mbox_trypost: mbox %p msg %p\n",
-                            (void *)mbox, (void *)msg));
+                            (void *)mb, (void *)msg));
 
-    if (!ext_sys_arch_sem_try(&mbox->write_sem)) {
+    if (!ext_sys_arch_sem_try(&mb->write_sem)) {
         return ERR_MEM;
     }
 
-    sys_arch_sem_wait(&mbox->mutex, 0);
-    mbox->msgs[mbox->last % SYS_MBOX_SIZE] = msg;
-    mbox->last++;
-    sys_sem_signal(&mbox->mutex);
+    KIRQL prev_irql = KeRaiseIrqlToSynchLevel();
```

Comment:
Is it necessary to raise the IRQL all the way up to sync level (=28), since that will block almost all other interrupts in the system. On the xbox, the NIC uses IRQ=4, which `HalGetInterruptVector` will convert to IRQL=22. Would that be enough here?

Reply:
Actually that was supposed to be `KeRaiseIrqlToDpcLevel`, I had increased it during debugging and apparently forgot to change it back.

Reply:
Happens to the best of us.